File: packages/backend/src/services/ai.ts
import OpenAI from 'openai';
import { fileTools, executeFileTool } from './fileTools';
import { getSystemPrompt } from '../prompts/system';

interface Message {
  role: 'user' | 'assistant';
  content: string;
}

/**
 * AI 服务类
 * 功能：与 DeepSeek API 交互，处理对话和工具调用
 * 输入：API key 和模型名称
 * 输出：对话结果和日志创建状态
 */
export class AIService {
  private client: OpenAI;
  private model: string;
  private logCreated = false;

  constructor(apiKey: string, model = 'deepseek-chat') {
    this.client = new OpenAI({
      apiKey,
      baseURL: 'https://api.deepseek.com',
    });
    this.model = model;
  }

  /**
   * 与 AI 对话（流式）
   * 输入：用户消息、历史对话记录、流式回调函数
   * 输出：通过回调函数流式返回内容，返回是否创建了日志的状态
   */
  async chatStream(
    userMessage: string,
    history: Message[] = [],
    onChunk: (chunk: string) => void
  ): Promise<{ logCreated: boolean }> {
    this.logCreated = false;

    const today = new Date().toISOString().split('T')[0];
    const [year, month] = today.split('-');
    const todayLogPath = `logs/${year}/${month}/${today}.md`;
    const systemPrompt = getSystemPrompt(today, todayLogPath);

    const messages: OpenAI.Chat.ChatCompletionMessageParam[] = [
      { role: 'system', content: systemPrompt },
      ...history.map(msg => ({
        role: msg.role as 'user' | 'assistant',
        content: msg.content
      })),
      { role: 'user', content: userMessage }
    ];

    try {
      let maxRounds = 5;

      while (maxRounds-- > 0) {
        const stream = await this.client.chat.completions.create({
          model: this.model,
          messages,
          tools: fileTools,
          stream: true,
        });

        let fullContent = '';
        let toolCalls: OpenAI.Chat.ChatCompletionMessageToolCall[] = [];
        const toolCallsMap = new Map<number, { id: string; name: string; arguments: string }>();

        for await (const chunk of stream) {
          const delta = chunk.choices[0]?.delta;
          if (!delta) continue;

          // 处理文本内容
          if (delta.content) {
            fullContent += delta.content;
            onChunk(delta.content);
          }

          // 处理工具调用
          if (delta.tool_calls) {
            for (const toolCall of delta.tool_calls) {
              const index = toolCall.index;
              if (!toolCallsMap.has(index)) {
                toolCallsMap.set(index, {
                  id: toolCall.id || '',
                  name: toolCall.function?.name || '',
                  arguments: ''
                });
              }
              const existing = toolCallsMap.get(index)!;
              if (toolCall.id) existing.id = toolCall.id;
              if (toolCall.function?.name) existing.name = toolCall.function.name;
              if (toolCall.function?.arguments) existing.arguments += toolCall.function.arguments;
            }
          }
        }

        // 转换工具调用
        toolCalls = Array.from(toolCallsMap.entries()).map(([index, call]) => ({
          id: call.id,
          type: 'function' as const,
          function: {
            name: call.name,
            arguments: call.arguments
          }
        }));

        // 如果没有工具调用，返回结果
        if (toolCalls.length === 0) {
          if (!fullContent.trim()) {
            onChunk('抱歉老大，我暂时没有获取到有效回复，请稍后再试。');
          }
          return { logCreated: this.logCreated };
        }

        // 添加助手消息到历史
        messages.push({
          role: 'assistant',
          content: fullContent || null,
          tool_calls: toolCalls
        });

        console.log('🔧 工具调用:', toolCalls.map(c => ({ name: c.function.name, args: c.function.arguments })));

        // 执行工具调用
        for (const toolCall of toolCalls) {
          const args = JSON.parse(toolCall.function.arguments);
          const toolResult = await executeFileTool(toolCall.function.name, args);
          console.log(`✅ ${toolCall.function.name} 结果:`, toolResult);

          if (toolCall.function.name === 'create_md_file' || toolCall.function.name === 'edit_md_file') {
            this.logCreated = true;
          }

          messages.push({
            role: 'tool',
            tool_call_id: toolCall.id,
            content: toolResult
          });
        }
      }

      onChunk('抱歉老大，处理请求超出了最大轮次限制。');
      return { logCreated: this.logCreated };
    } catch (error) {
      console.error('AIService.chatStream: 调用失败', error);
      onChunk('抱歉老大，我暂时无法连接到 AI 服务，请稍后再试。');
      return { logCreated: false };
    }
  }
}

File: packages/backend-py/src/services/ai.py
"""
AI 服务模块
处理与 Google Gemini API 的交互
"""

import os
from typing import Any

import google.generativeai as genai

from ..prompts.system import get_system_prompt
from .file_tools import FILE_TOOLS, execute_file_tool


class AIService:
    """AI 服务类"""

    def __init__(self, api_key: str, model: str = "gemini-pro"):
        """
        初始化 AI 服务

        参数:
            api_key: Google API Key
            model: 模型名称
        """
        # 配置代理
        proxy = os.getenv("HTTPS_PROXY") or os.getenv("HTTP_PROXY")
        if proxy:
            os.environ["HTTP_PROXY"] = proxy
            os.environ["HTTPS_PROXY"] = proxy
            print(f"使用代理: {proxy}")

        genai.configure(api_key=api_key)
        self.model_name = model
        self.log_created = False

    async def chat(self, user_message: str, history: list[dict[str, str]] = None) -> dict[str, Any]:
        """
        处理聊天请求

        参数:
            user_message: 用户消息
            history: 聊天历史 [{"role": "user/assistant", "content": "..."}]

        返回:
            Dict: {"message": "回复内容", "logCreated": bool}
        """
        if history is None:
            history = []

        self.log_created = False

        # 获取今天的日期和路径
        from datetime import date

        today = date.today().isoformat()
        year, month = today.split("-")[:2]
        today_log_path = f"logs/{year}/{month}/{today}.md"

        # 获取系统提示
        system_prompt = get_system_prompt(today, today_log_path)

        # 转换历史记录格式
        chat_history = []
        for msg in history:
            role = "user" if msg["role"] == "user" else "model"
            chat_history.append({"role": role, "parts": [msg["content"]]})

        try:
            # 创建模型实例
            model = genai.GenerativeModel(
                model_name=self.model_name,
                tools=FILE_TOOLS,
                system_instruction=system_prompt,
            )

            # 开始聊天
            chat = model.start_chat(history=chat_history)
            response = chat.send_message(user_message)

            # 处理工具调用（最多5轮）
            max_rounds = 5
            while max_rounds > 0 and response.candidates[0].content.parts:
                function_calls = []

                # 提取函数调用
                for part in response.candidates[0].content.parts:
                    if hasattr(part, "function_call") and part.function_call:
                        function_calls.append(part.function_call)

                if not function_calls:
                    break

                print(
                    f"🔧 工具调用: {[{'name': fc.name, 'args': dict(fc.args)} for fc in function_calls]}"
                )

                # 执行工具
                function_responses = []
                for fc in function_calls:
                    tool_result = await execute_file_tool(fc.name, dict(fc.args))
                    print(f"✅ {fc.name} 结果: {tool_result}")

                    # 检查是否创建/更新了日志
                    if fc.name in ["create_md_file", "edit_md_file"]:
                        self.log_created = True

                    function_responses.append(
                        genai.protos.Part(
                            function_response=genai.protos.FunctionResponse(
                                name=fc.name, response={"result": tool_result}
                            )
                        )
                    )

                # 发送工具响应
                response = chat.send_message(function_responses)
                max_rounds -= 1

            # 获取最终文本
            text = response.text if response.text else ""

            if not text.strip():
                print("AIService.chat: empty response")
                return {
                    "message": "抱歉老大，我暂时没有获取到有效回复，请稍后再试。",
                    "logCreated": False,
                }

            return {"message": text, "logCreated": self.log_created}

        except Exception as e:
            print(f"AIService.chat: 调用失败 {type(e).__name__}: {str(e)}")
            return {
                "message": "抱歉老大，我暂时无法连接到 AI 服务，请稍后再试。",
                "logCreated": False,
            }
